executor/executor.go

```diff
@@ -2318,3 +2331,470 @@ func isWeakConsistencyRead(ctx sessionctx.Context, node ast.Node) bool {
+
+	trySaveErr := func(err error) {
+		if setNewErr := w.e.hasError.CompareAndSwap(false, true); setNewErr {
+			w.e.err = err
+		}
+	}
+
+	se, err := w.e.base().getSysSession()
+	se.GetSessionVars().OptimizerUseInvisibleIndexes = true
+	if err != nil {
+		trySaveErr(err)
+		return
+	}
+	defer func() {
+		se.GetSessionVars().OptimizerUseInvisibleIndexes = false
+		w.e.base().releaseSysSession(ctx, se)
+	}()
+
+	var pkCols []string
+	var pkTypes []*types.FieldType
+	var primaryIdx *model.IndexInfo
+	switch {
+	case w.e.table.Meta().IsCommonHandle:
+		pkColsInfo := w.e.table.Meta().GetPrimaryKey().Columns
+		for _, colInfo := range pkColsInfo {
+			colStr := colInfo.Name.O
+			pkCols = append(pkCols, colStr)
+		}
+		primaryIdx = tables.FindPrimaryIndex(w.table.Meta())
+	case w.e.table.Meta().PKIsHandle:
+		pkCols = append(pkCols, w.e.table.Meta().GetPkName().O)
+		primaryIdx = tables.FindPrimaryIndex(w.table.Meta())
+	default: // support decoding _tidb_rowid.
+		pkCols = append(pkCols, model.ExtraHandleName.O)
+	}
+
+	// CheckSum of (handle + index columns).
+	var md5HandleAndIndexCol strings.Builder
+	md5HandleAndIndexCol.WriteString("crc32(md5(concat(")
+	for _, col := range pkCols {
+		md5HandleAndIndexCol.WriteString(col)
+		md5HandleAndIndexCol.WriteString(", ")
+	}
+	for offset, col := range idxInfo.Columns {
+		md5HandleAndIndexCol.WriteString("ifnull(")
+		tblCol := w.table.Meta().Columns[col.Offset]
+		if tblCol.IsGenerated() && !tblCol.GeneratedStored {
+			md5HandleAndIndexCol.WriteString(tblCol.GeneratedExprString)
+		} else {
+			md5HandleAndIndexCol.WriteString(col.Name.O)
+		}
+		md5HandleAndIndexCol.WriteString(", 0x1)")
+		if offset != len(idxInfo.Columns)-1 {
+			md5HandleAndIndexCol.WriteString(", ")
+		}
+	}
+	md5HandleAndIndexCol.WriteString(")))")
+
+	// Used to group by and order.
+	var md5Handle strings.Builder
+	md5Handle.WriteString("crc32(md5(concat(")
+	for i, col := range pkCols {
+		md5Handle.WriteString(col)
+		if i != len(pkCols)-1 {
+			md5Handle.WriteString(", ")
+		}
+	}
+	md5Handle.WriteString(")))")
+
+	handleColumnField := strings.Join(pkCols, ", ")
+	var indexColumnField strings.Builder
+	for offset, col := range idxInfo.Columns {
+		indexColumnField.WriteString(col.Name.O)
+		if offset != len(idxInfo.Columns)-1 {
+			indexColumnField.WriteString(", ")
+		}
+	}
+
+	tableRowCntToCheck := int64(0)
+
+	offset := 0
+	mod := 1
+	meetError := false
+
+	lookupCheckThreshold := int64(100)
+	checkOnce := false
+
+	if w.e.ctx.GetSessionVars().SnapshotTS != 0 {
+		_, err = se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, "set @@tidb_snapshot = %?", w.e.ctx.GetSessionVars().SnapshotTS)
+		if err != nil {
+			trySaveErr(err)
+			return
+		}
+		defer func() {
+			_, err = se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, "set @@tidb_snapshot = ''")
+			if err != nil {
+				trySaveErr(err)
+				return
+			}
+		}()
+	}
+	_, err = se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, "begin")
```

Comment:
We should rollback or commit the transaction before returning it to the session pool.